Fix `Optimize1qGatesDecomposition` length heuristic

ecpeterson opened the pull request:
<!--
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

The decision in `Optimize1qGatesDecomposition` whether to use the re-synthesized gate string or the original gate string is decided purely by length: we prefer whichever of the new string and the old string is shorter. This is a rough approximation to expected infidelity cost of the sequence _when both strings are of native gates_. An input string with non-native gates might be unnaturally short, hence persist into the output — a bug.

### Details and comments

Before:
```
In [2]: from qiskit import QuantumCircuit
   ...: from qiskit.transpiler.passes import Optimize1qGatesDecomposition
   ...: import numpy as np
   ...: 
   ...: c = QuantumCircuit(1)
   ...: c.h(0)
   ...: c.ry(np.pi/2, 0)
   ...: print(Optimize1qGatesDecomposition(['sx', 'rz'])(c))
     ┌───┐┌─────────┐
q_0: ┤ H ├┤ RY(π/2) ├
     └───┘└─────────┘
```

After:
```
In [1]: from qiskit import QuantumCircuit
   ...: from qiskit.transpiler.passes import Optimize1qGatesDecomposition
   ...: import numpy as np
   ...: 
   ...: c = QuantumCircuit(1)
   ...: c.h(0)
   ...: c.ry(np.pi/2, 0)
   ...: print(Optimize1qGatesDecomposition(['sx', 'rz'])(c))
     ┌────┐┌────┐
q_0: ┤ √X ├┤ √X ├
     └────┘└────┘
```

I also removed some special casing on `U3` and single gates that seemed to overlap with this oversight. Happy to reverse that.

ajavadia commented on releasenotes/notes/fixed-bug-in-Optimize1qGatesDecomposition-skipping-short-sequences-044a64740bf414a7.yaml:
```diff
@@ -0,0 +1,65 @@
+---
```
This file you have to delete all the boilerplate and keep only the part that is relevant to this PR.

ecpeterson replied:
haha, serves me right for not looking at the diff

ajavadia commented on qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py:
```diff
@@ -102,7 +83,7 @@ def run(self, dag):
             if new_circs:
                 new_circ = min(new_circs, key=len)
-                if len(run) > len(new_circ) or (single_u3 and new_circ.data[0][0].name != "u3"):
+                if any([g.name not in self.basis for g in run]) or len(run) > len(new_circ):
```
nit
```suggestion
                if any(g.name not in self.basis for g in run) or len(run) > len(new_circ):
```

ajavadia reviewed (commented):
looks good, two small comments

levbishop commented:
I wonder if we can't completely remove the check if the new run is shorter. With the changes in #5827 I'm hoping there should be no remaining cases where the resynthesized sequence is ever non-optimal (and if any do remain, I consider it a bug).

ecpeterson commented:
I added a branch for a warning, rather than an error. It would be obnoxious for a synthesis flaw to generate suboptimal programs, but probably worse to then prevent a user from running at all.

mtreinish commented on qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py:
```diff
@@ -102,7 +84,13 @@ def run(self, dag):
+                                  f"but the original was native and the new "
+                                  f"value is longer.  This indicates an "
+                                  f"efficiency bug in synthesis.  Please "
+                                  f"report it!")
```
Maybe add the link to the bug tracker:

```suggestion
                                  f"report it by opening an issue here:"
                                  f"https://github.com/Qiskit/qiskit-terra/issues/new/choose")
```

mtreinish replied:
Also maybe we want to set a stacklevel so the warning doesn't show the warn line, maybe `stacklevel=2` so it shows the `optimize_1q_decomposition.run()` call).

mtreinish commented on releasenotes/notes/fixed-bug-in-Optimize1qGatesDecomposition-skipping-short-sequences-044a64740bf414a7.yaml:
```diff
@@ -0,0 +1,5 @@
+---
+fixes:
+  - |
+    Fixes a bug in `Optimize1qGatesDecomposition` previously causing certain
```
```suggestion
    Fixes a bug in :func:`~qiskit.transpiler.passes.Optimize1qGatesDecomposition` previously causing certain
```

mtreinish commented on qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py:
```diff
@@ -102,7 +84,13 @@ def run(self, dag):
             if new_circs:
                 new_circ = min(new_circs, key=len)
-                if len(run) > len(new_circ) or (single_u3 and new_circ.data[0][0].name != "u3"):
+                if all(g.name in self.basis for g in run) and len(run) < len(new_circ):
```
Do you have an idea on what the runtime cost of this additional check is? If it's noticeable I'd rather not add this to just raise a warning because this will be called quite a lot during a transpile call.

ecpeterson replied:
I don't have data, but I imagine it's negligible relative to the nonnegotiable cost of constructing `new_circ`: `self.basis` tends not to be super large, and `decomposer` has to read all of `run` anyhow. I'll do a small amount of with/without benchmarking.

mtreinish reviewed (commented):
Overall this LGTM thanks for fixing this. I agree all that logic is out of date with #5827 now since the issues that single gate case were there for have been fixed. This will also probably improve performance a bit since we're not going to do an identity check on each standalone 1q gate. Just a few nits inline, but then I think this good to go.

Also can you update the release note to say you've also fixed #6473 because I believe this will fix that issue too. (basically just add a new bullet point to the release note yaml for the second fix).

mtreinish commented:
Oh I didn't see the test failures before reviewing and just assumed they all passed. We probably should get to the bottom of those. Most I think are fine and we just need to update the tests like `test_short_string`: https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=28674&view=logs&j=8eacdd59-c2e8-5617-75a5-8ed7c854a78f&t=5df005f4-d5f4-5d2d-0307-5b24b15488fc&l=12271 we can just update that to be `X` as the expected.

But `test_euler_decomposition_worse` might be an issue because it looks like the output is worse than what we had before:
https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=28674&view=logs&j=8eacdd59-c2e8-5617-75a5-8ed7c854a78f&t=5df005f4-d5f4-5d2d-0307-5b24b15488fc&l=12338

levbishop commented:
> But `test_euler_decomposition_worse` might be an issue because it looks like the output is worse than what we had before

Yeah this is tricky. The Euler Z(phi)X(theta)Z(lambda) decomposition should find an optimal expansion for 0<=theta<=pi. The input circuit has theta=-pi/2 outside that range, so the Euler decomposer includes the extra Z(). It wouldn't be crazy to  change the Euler decomposer to check in case the negative theta := theta-2pi allows such a simplification and emit that if its fewer gates.

On the other hand, both of these solutions are optimal in terms of X() gates which is what actually cost something for many hardware implementations.

ecpeterson commented:
Some of the test problems were legitimate: the slot `Optimize1qGatesDecomposition.basis` apparently stores _functions which map into_ the basis described by the `basis` init kwarg — and that kwarg itself gets tossed. So, `... in self.basis` didn't do what I expected (or probably what people reading this PR expected).

Fixed that, updated some of the reference circuits in the test suite, but this still isn't quite good to go: `test.python.compiler.test_transpiler.TestTranspile.test_transpile_calibrated_nonbasis_gate_on_diff_qubit` checks that non-native gates do / don't get rewritten if they don't / do have calibration definitions, but this change makes rewriting more aggressive. Stay tuned.

ajavadia commented on qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py:
```diff
@@ -69,40 +67,46 @@ def run(self, dag):
+                # if we're outside of the basis set, we're obligated to logically decompose.
+                # if we're outside of the set of gates for which we have physical definitions,
+                #    then we _try_ to decompose, using the results if we see improvement.
+                if uncalibrated_and_not_basis_p or (uncalibrated_p and len(run) >= len(new_circ)):
```
i don't really understand what the "_p" means in these names but the overall logic looks good to me!

ecpeterson replied:
"This returns / is a boolean", with p short for predicate. I did it out of mindless old habit, though now that you mention it I recognize that it's idiosyncratic. I don't mind dropping the `_p` if you'd prefer.

mtreinish commented:
I ran the quantum volume benchmarks from the asv benchmark suite on this just now and it causes a roughly 30% run time performance regression:

```
```

My assumption is that this is caused by all the times we're iterating over the gates in each run to check for calibrations now. I haven't profiled it yet but I can do that on monday. But until we get to the bottom of this I've removed the automerge label to prevent mergify from automerging this.

levbishop commented:
It looks like `test_euler_decomposition_worse` is now passing but I'm missing what changed to make it pass?

ajavadia commented:
> My assumption is that this is caused by all the times we're iterating over the gates in each run to check for calibrations now

The common case is that the special cals don't exist, so I agree these individuals checks shouldn't happen all the time.  So I would do a check upfront to see if `dag.calibrations is not None`. The slow path should only be taken in such a case.

ecpeterson commented:
@levbishop : I'm pretty sure fec5454 / https://github.com/Qiskit/qiskit-terra/pull/6553#issuecomment-859190279 is the change that fixed that failure.

ecpeterson commented:
@ajavadia There's a test that specifically checks that cal-preservation only covers the qubits which have cals, so tossing everything into `basis_gates` won't work. I'll do the early short-circuiting option.

ecpeterson commented:
The actual culprit is a whole bunch of new calls to `_decompose`, as well as a bunch of calls to `substitute_node_with_dag`. These are surely consequences of deleting that big block of guards that abstained from rewriting single gates (except for U3 (sometimes)).

I put some of the logic from that block back: if U3 is in the target basis and this is a run of length 1, then we know that the U3 decomposer isn't going to do better than what we've already got and it's safe to skip. I find it kind of unpalatable to have to state that special case, especially since there are similar special cases for U2 and U1 (not previously included, and I'm not inclined to include them now), but I think it's the sanest thing in the current set-up.

Waiting for tests to pass locally.

ecpeterson commented:
Not sure what's up with the test suite now: it times out while running sphinx on the tutorials.

ajavadia commented:
Yeah I don't love this new logic. It's basically saying if you find a single U3 and that is native, let it stay. Except when it is identity (remove it), or u1/u2 are native (hope the U3 will simplify).

Yes the old code was faster but it also had all these if/else conditionals that make the code work for one set of basis and not others. For one, U3Gate is being deprecated in favor of UGate. U1 is similarly being deprecated in favor of PhaseGate, and RZGate won't be recognized in this logic. 

I think what we really want is for the decomposer to have a cost for each gate, and e.g. recognize that `RZ.SX.RZ` is cheaper than `U`, even though U is shorter as a sequence. I don't mind if the code becomes slower due to these extra checks, as long as it can make the circuit better. We can put it as an option in level 3 or something. 

Anyway for now the rewrite decision is based on length of sequence. I think the U1/U2 logic is not so useful anymore as these used to be IBM backend basis, but are not anymore. So I'd just remove that part. And use U instead of (or in addition to) U3.

mtreinish commented:
> Yeah I don't love this new logic. It's basically saying if you find a single U3 and that is native, let it stay. Except when it is identity (remove it), or u1/u2 are native (hope the U3 will simplify).
> 
> Yes the old code was faster but it also had all these if/else conditionals that make the code work for one set of basis and not others. For one, U3Gate is being deprecated in favor of UGate. U1 is similarly being deprecated in favor of PhaseGate, and RZGate won't be recognized in this logic.

Don't underestimate the importance of the speed for this pass, when routing isn't a concern (so narrow circuits with a modest number of qubits or without a coupling map set) this pass is typically the bottleneck of a transpile. It also gets called many times as it's in pretty much every pass manager's optimization loop so a small difference in performance can lead to noticeable slowdowns.

> 
> I think what we really want is for the decomposer to have a cost for each gate, and e.g. recognize that `RZ.SX.RZ` is cheaper than `U`, even though U is shorter as a sequence. I don't mind if the code becomes slower due to these extra checks, as long as it can make the circuit better. We can put it as an option in level 3 or something.

I think having a better heuristic for selection makes sense, we used length here as the best generally available one but making it configurable based on the availability of error rates or other properties makes sense as a future enhancement. I think adding a real instruction cost field for this kind of thing to the backend/target in the next iteration of the backend class (see #5885 for where I'm collecting ideas) makes a lot of sense, just so we have a priority list set by the backend for these kind of decisions.

> 
> Anyway for now the rewrite decision is based on length of sequence. I think the U1/U2 logic is not so useful anymore as these used to be IBM backend basis, but are not anymore. So I'd just remove that part. And use U instead of (or in addition to) U3.

This is actually a necessary component of this pass for analysis of circuits on older backends. As this pass is the default in all the preset pass managers if you're running the transpiler with a u3/u2/u1 basis as the target (which people still do to validate old runs, etc) having the lowering to u2 and u1 is important because that was the key difference in behavior between this pass and the old optimize 1q gates for the u3/u2/u1 basis (obviously the old pass didn't work for any other basis). We discussed this at great length before making this pass the default everywhere (#5468 was the tracking issue but the discussion was light in the actual issue).

ajavadia commented:
> As this pass is the default in all the preset pass managers if you're running the transpiler with a u3/u2/u1 basis as the target (which people still do to validate old runs, etc) having the lowering to u2 and u1 is important because that was the key difference in behavior between this pass and the old optimize 1q gates for the u3/u2/u1 basis

Well at a minimum it should work for the new basis. Can we add a test that U(0,0,pi/6) gets optimized to Phase(pi/6) or RZ(pi/6) (whichever is in the basis). Far more people now use the new basis.

ecpeterson commented:
@ajavadia: I added the couple of indicated tests in e804e49 . Let me know if that isn't what you were asking for.

ajavadia commented:
Is the speed back to where it was now? If so let's merge

ecpeterson commented:
In my local tests, the speed is back to where it was — but I didn't figure out how to run `asv`. If @mtreinish sees the same thing, I'm game to merge.

levbishop commented:
> @levbishop : I'm pretty sure [fec5454](https://github.com/Qiskit/qiskit-terra/commit/fec545497ac6d3faa32884bdbf56d192f472df00) / [#6553 (comment)](https://github.com/Qiskit/qiskit-terra/pull/6553#issuecomment-859190279) is the change that fixed that failure.

It didn't really "fix" it though, right? In the sense that it generates the warning but then falls back to the original circuit. Seems a bit unfriendly to put this out with the warning enabled and known warnings like this. (Aside, shouldn't the test suite be set up to error/fail on unexpected warnings?)

ecpeterson commented:
Oh, I thought it had resolved it, and that there wasn't a warning! I'll look first thing tomorrow.

For what it's worth, my local test runner _does_ fail on warnings. It's currently mad about some cvxpy code that instantiates some numpy matrices using an outdated dtype synonym. I'd hope that if it's upset about that, it would also be upset about warnings within QISKit — but I will double check.

levbishop commented:
#6392 made DeprecationWarning an error, except for a list of known depreciations, with the idea we gradually shrink the list to zero.

ecpeterson commented:
@levbishop I've pushed a change that commutes the left-hand rotation through in the case `lambda = ± pi`, which resolves the warning in that test case.

This maybe isn't the best fix: rather than normalizing a generic lambda to lie in the open interval (-pi, pi), it might be preferable to use [0, pi). Depends on how canonical one wants these canonical decompositions to really be.

mtreinish commented:
> In my local tests, the speed is back to where it was — but I didn't figure out how to run `asv`. If @mtreinish sees the same thing, I'm game to merge.

I'm still seeing a performance regresssion:

```
Benchmarks that have got worse:

       before           after         ratio
     [a2e5b9d9]       [93c2e751]
     <main>       <ecpeterson-bugfix/nonnative-1q-heuristic>
!         24.0±0s           failed      n/a  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(14, 'synthesis')
+        228±10ms         341±20ms     1.49  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(8, 'translator')
+        720±50ms       1.05±0.09s     1.46  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(14, 'translator')
+     14.6±0.02ms       19.1±0.1ms     1.31  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(2, 'translator')
+        76.7±2ms         97.8±8ms     1.28  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(5, 'translator')
+      2.60±0.02s       3.30±0.01s     1.27  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(27, 'translator')
+      1.63±0.01s        2.03±0.2s     1.24  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(20, 'translator')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

levbishop commented:
> @levbishop I've pushed a change that commutes the left-hand rotation through in the case `lambda = ± pi`, which resolves the warning in that test case.

This covers the case where lambda could be 0 but not the corresponding case for phi:
```python
In [6]: In [2]: from qiskit import QuantumCircuit
   ...:    ...: from qiskit.transpiler.passes import Optimize1qGatesDecomposition
   ...:    ...: import numpy as np
   ...:    ...:
   ...:    ...: c = QuantumCircuit(1)
   ...:    ...: c.rz(0.13, 0)
   ...:    ...: c.ry(-0.14, 0)
   ...:    ...: print(Optimize1qGatesDecomposition(['ry', 'rz'])(c))
/Users/lsbishop/ghpub/terra/master/qiskit/transpiler/basepasses.py:120: UserWarning: Resynthesized [<qiskit.dagcircuit.dagnode.DAGNode object at 0x7fe79a25a520>, <qiskit.dagcircuit.dagnode.DAGNode object at 0x7fe79a25a0a0>] and got global phase: π
      ┌─────────────┐┌──────────┐┌────────┐
qr_0: ┤ RZ(-3.0116) ├┤ RY(0.14) ├┤ RZ(-π) ├
      └─────────────┘└──────────┘└────────┘, but the original was native and the new value is longer.  This indicates an efficiency bug in synthesis.  Please report it by opening an issue here: https://github.com/Qiskit/qiskit-terra/issues/new/choose
  result = self.run(circuit_to_dag(circuit))
     ┌──────────┐┌───────────┐
q_0: ┤ RZ(0.13) ├┤ RY(-0.14) ├
     └──────────┘└───────────┘
```

and there are versions of this problem for other expansions such as ZSX:
```python
In [29]: from qiskit import QuantumCircuit
    ...: from qiskit.transpiler.passes import Optimize1qGatesDecomposition
    ...: import numpy as np
    ...:
    ...: c = QuantumCircuit(1)
    ...: c.rz(0.3, 0)
    ...: c.sx(0)
    ...: c.rz(0.2, 0)
    ...: c.sx(0)
    ...: #c.rz(0.1, 0)
    ...: print(Optimize1qGatesDecomposition(['sx', 'rz'])(c))
/Users/lsbishop/ghpub/terra/master/qiskit/transpiler/basepasses.py:120: UserWarning: Resynthesized [<qiskit.dagcircuit.dagnode.DAGNode object at 0x7fe79a43a5e0>, <qiskit.dagcircuit.dagnode.DAGNode object at 0x7fe79a3f9b80>, <qiskit.dagcircuit.dagnode.DAGNode object at 0x7fe79a3f9ac0>, <qiskit.dagcircuit.dagnode.DAGNode object at 0x7fe79a3f9a00>] and got global phase: 0
      ┌─────────────┐┌────┐┌──────────┐┌────┐┌────────┐
qr_0: ┤ RZ(-2.8416) ├┤ √X ├┤ RZ(-0.2) ├┤ √X ├┤ RZ(-π) ├
      └─────────────┘└────┘└──────────┘└────┘└────────┘, but the original was native and the new value is longer.  This indicates an efficiency bug in synthesis.  Please report it by opening an issue here: https://github.com/Qiskit/qiskit-terra/issues/new/choose
  result = self.run(circuit_to_dag(circuit))
     ┌─────────┐┌────┐┌─────────┐┌────┐
q_0: ┤ RZ(0.3) ├┤ √X ├┤ RZ(0.2) ├┤ √X ├
     └─────────┘└────┘└─────────┘└────┘
```

> This maybe isn't the best fix: rather than normalizing a generic lambda to lie in the open interval (-pi, pi), it might be preferable to use [0, pi). Depends on how canonical one wants these canonical decompositions to really be.

ecpeterson commented:
The performance regression appears to be from extra calls to `circuit_to_dag`, owing to the weak inequality on this line: https://github.com/Qiskit/qiskit-terra/pull/6553/files#diff-85095a0baf57c8f5835b68b9e61d4acf6eb425e1f918a313f22b8ec9106c3e76R129 . If I change it to a strict inequality, the `avs` benchmarks largely snap back to their normal speed (though they display _significant_ variance over runs, so I'm not super confident in this). Meanwhile, the tests get crabby because they expect this pass to lower `u3` to `u2` and `u1` where possible, which the weak inequality doesn't catch.

I guess since it's already cheating to use circuit gate count as a proxy for runtime fidelity, I then don't have a good argument that there shouldn't be further special-casing logic handling the reduction of a few gates we care about. 2f7f3393 implements this compromise.

ecpeterson commented:
> This covers the case where lambda could be 0 but not the corresponding case for phi, and there are versions of this problem for other expansions such as ZSX.

Yuck, OK.

levbishop commented on qiskit/quantum_info/synthesis/one_qubit_decompose.py:
```diff
@@ -416,32 +422,46 @@ def _circuit_psx_gen(theta, phi, lam, phase, atol, pfun, xfun, xpifun=None):
         if abs(theta - np.pi) < atol:
             circuit.global_phase += lam
             phi, lam = phi - lam, 0
+        if abs(_mod_2pi(lam + np.pi)) < atol:
+            lam, theta, phi = 0, -theta, phi + lam
+            circuit.global_phase -= theta
+        if abs(_mod_2pi(phi)) < atol:
+            lam, theta, phi = lam - np.pi, -theta, phi + np.pi
+            circuit.global_phase -= theta
```
maybe clearer to move this after the shift? Probably also better to have a single if:
```python
if abs(_mod_2pi(phi-pi))<atol or abs(_mod_2pi(lam-pi))<atol:
    theta, phi, lam = -theta, phi+pi, lam+pi
```

ecpeterson replied:
Agreed about the `if`. I found this clearer before the shift: I spent a while confused about why these extra RZ(pi)s weren't flipping the SXes to SX^*es, but when that chunk is written as an RY I see it straight away.

levbishop commented on qiskit/quantum_info/synthesis/one_qubit_decompose.py:
```diff
@@ -287,28 +287,30 @@ def _circuit_zyz(theta, phi, lam, phase, simplify=True, atol=DEFAULT_ATOL):
         if abs(_mod_2pi(lam + np.pi)) < atol:
             lam, theta, phi = 0, -theta, lam + phi
-        if abs(lam) > atol:
+        if abs(_mod_2pi(phi + np.pi)) < atol:
```
maybe better to make it a single if:
```python
if abs(_mod_2pi(lam-np.pi))<atol or abs(_mod_2pi(phi-np.pi))<atol:
   theta, phi, lam = -theta, phi+pi, lam+pi
```
Then we could also make the if a bit smarter to cover different concepts of optimality:
eg add a flag `allow_non_canonical` (default `True`) like we have `simplify`, and the check can be:
```python
if allow_non_canonical and (abs(_mod_2pi(lam-np.pi))<atol or abs(_mod_2pi(phi-np.pi))<atol):
   theta, phi, lam = -theta, phi+pi, lam+pi
```
which I think covers the 2 different behaviours one might desire from the decomposer.

ecpeterson replied:
This reminds me that I haven't been consistent about guarding with `simplify`.

levbishop replied:
Before I refactored it was wrapping "most" of the elisions. I refactored it to set atol=-1 and thus wrap all elisions

ecpeterson commented on qiskit/quantum_info/synthesis/one_qubit_decompose.py:
```diff
@@ -287,28 +287,30 @@ def _circuit_zyz(theta, phi, lam, phase, simplify=True, atol=DEFAULT_ATOL):
-            if abs(tot) > atol:
-                circuit._append(RZGate(tot), [qr[0]], [])
-                gphase += tot / 2
+        if abs(_mod_2pi(theta)) < atol:
```
I'll trade this for a comment that that's the expectation, then. `._params` isn't consistent about the range expectations, and while the individual `._params_...` are, they don't announce it.

ecpeterson commented on qiskit/quantum_info/synthesis/one_qubit_decompose.py:
```diff
@@ -287,28 +287,30 @@ def _circuit_zyz(theta, phi, lam, phase, simplify=True, atol=DEFAULT_ATOL):
-                circuit._append(RZGate(tot), [qr[0]], [])
-                gphase += tot / 2
+        if abs(_mod_2pi(theta)) < atol:
+            lam, phi = lam + phi, 0
+            if abs(_mod_2pi(lam)) > atol:
+                circuit._append(RZGate(lam), [qr[0]], [])
+                gphase += lam / 2
```
These changes were intentional, but maybe wrong-headed. I got bitten in an earlier version of this commit by wrapping everything to (-pi, pi], in both the conditionals and the actual parameters sent to the gates, which caused some phase-tracking violations. This looked like an instance of the same thing, but in the original source, so I changed it to fit the rest.

But I'm not sure I understand the complaint. Is it that I'm not passing `atol` to `_mod_2pi`? That's an oversight which I should fix.

levbishop replied:
passing atol to _mod_2pi in the conditional won't do anything. but previously it would make a difference since it wrapped the parameter

ecpeterson replied:
Is there another input range reason why the the original code was never in danger of trading RZ(pi + pi) for RZ(0) and slipping global phase?

levbishop replied:
i'm not sure actually. i thought i had pretty complete test coverage so either there is some guarantee or my tests not as complete as expected...

ecpeterson replied:
Ah: the phase normalization that's already here fixes one of the diagonal entries to 1, and that's enough to prevent phase slipping.

ecpeterson commented on qiskit/quantum_info/synthesis/one_qubit_decompose.py:
```diff
@@ -318,28 +320,30 @@ def _circuit_zxz(theta, phi, lam, phase, simplify=True, atol=DEFAULT_ATOL):
             circuit._append(RZGate(phi), [qr[0]], [])
-        circuit.global_phase = gphase
+        circuit.global_phase = _mod_2pi(gphase)
         return circuit
```
Rather than duplicate this work, I'm going to collapse these functions into one.

levbishop replied:
:+1: see also #6605

levbishop commented on qiskit/quantum_info/synthesis/one_qubit_decompose.py:
```diff
@@ -413,35 +375,45 @@ def _circuit_u(theta, phi, lam, phase, simplify=True, atol=DEFAULT_ATOL):
+        """
+        Generic X90, phase decomposition
+
+        NOTE: `pfun`, `xfun`, and `xpifun` are responsible for eliding gates
+              where appropriate (e.g., at angle value 0).
```
```suggestion
        NOTE: `pfun` is responsible for eliding gates where appropriate (e.g., at angle value 0).
```

levbishop reviewed (commented):
Approved modulo couple of suggested tiny tweaks around `_mod_2pi(..., atol)`. 

I kind of lost the thread regarding performance impacts, but I'm under the impression that the remaining impact is primarily due to `substitute_node_with_dag` being called more often due to more runs considered eligible for resynthesis? If so I guess #6302 will mitigate most of the impact when that lands. Either way, it looks this PR only calls `substitute_node_with_dag` in cases where it is in fact necessary (out of basis, new run is shorter, etc), so we have to do it and deal with the performance?

In reviewing this I got a few ideas for further simplifying one_qubit_decompose.py logic, and also maybe to improve the fidelity when `simplify` is true (I want to look into whether dropping some gates where the angle is less than `atol` maybe we should update the global phase based on the elision). I'll take a look at these things when I get back, no need for them to weigh down this PR.

levbishop commented:
It would probably be a good idea to add a couple of extra lines to `ANGEXP_ZYZ` and `ANGEXP_PSX` for  `TestOneQubitEulerSpecial` to exercise the new code paths for simplifying by flipping the sign of theta.

ecpeterson commented:
> ... `ANGEXP_ZYZ` and `ANGEXP_PSX` ...

Good idea. Done.

mtreinish commented:
I reran benchmarks just now, it looks like the performance regression from ealier has now been resolved and the performance is equivalent to the current `main` branch:

```
Benchmarks that have stayed the same:

       before           after         ratio
     [e9f06e45]       [2fdd628f]
     <main>           <ecpeterson-bugfix/nonnative-1q-heuristic>
           failed           failed      n/a  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(14, 'synthesis')
           failed           failed      n/a  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(20, 'synthesis')
           failed           failed      n/a  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(27, 'synthesis')
          5.83±1s          7.16±3s    ~1.23  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(8, 'synthesis')
          1.90±0s       2.23±0.04s    ~1.18  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(20, 'translator')
        3.60±0.2s       3.98±0.07s    ~1.10  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(27, 'translator')
       5.71±0.5ms       6.24±0.6ms     1.09  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(1, 'synthesis')
         33.7±3ms       35.1±0.5ms     1.04  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(3, 'translator')
         98.6±4ms          101±5ms     1.03  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(5, 'translator')
        720±100ms        739±500ms     1.03  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(5, 'synthesis')
          1.03±0s       1.05±0.06s     1.02  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(14, 'translator')
       23.5±0.2ms       23.8±0.2ms     1.02  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(2, 'synthesis')
         20.3±1ms      20.4±0.04ms     1.01  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(2, 'translator')
         344±20ms         319±20ms     0.93  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(8, 'translator')
       1.95±0.2ms      1.76±0.02ms    ~0.90  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(1, 'translator')
         130±80ms         106±60ms    ~0.82  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(3, 'synthesis')

BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

I'm a bit concerned about that variance, especially as I ran this on my local benchmark system which is doesn't have anything else running on it. But it doesn't look outside the norm for the benchmark (my guess is that the seeding for the qv circuit generation isn't correct or something like that).

ecpeterson commented:
I also see really strong variance across benchmarking runs, and it sounds worth looking into. I don't think the variance is something I introduced, but who knows what it might be covering up.

mtreinish commented on qiskit/quantum_info/synthesis/one_qubit_decompose.py:
```diff
@@ -281,91 +281,112 @@ def _params_u1x(mat):
+        k_gate=RZGate,
+        a_gate=RYGate,
+    ):
+        """
+        Installs the angles phi, theta, and lam into a KAK-type decomposition of the form
+        K(phi) . A(theta) . K(lam) , where K and A are an orthogonal pair drawn from RZGate, RYGate,
+        and RXGate.
+
+        Behavior flags:
+            `simplify` indicates whether gates should be elided / coalesced where possible.
+            `allow_non_canonical` indicates whether we are permitted to reverse the sign of the
+                middle parameter, theta, in the output.  When this and `simplify` are both enabled,
+                we take the opportunity to commute half-rotations in the outer gates past the middle
+                gate, which permits us to coalesce them at the cost of reversing the sign of theta.
+
+        NOTE: The input value of `theta` is expected to lie in [0, pi).
+        """
```
Since this is a private method and neither the docs builds or linters care about docstring formatting on this isn't a big issue. But I wonder if it would be better to restructure this in the normal docstrig format for the [napoleon plugin](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html) just to be consistent with the rest of the docs. If this were a built and published doc it wouldn't actually pass CI.

mtreinish reviewed (approved):
LGTM, one nit inline about docs formatting but it's not worth blocking over as it doesn't get published. I definitely appreciate all the comments added, it makes it easier to trace through things now.

levbishop reviewed (approved):
This was a marathon... thanks for sticking with it
